### ISSUE TITLE
Derivation on scala 3 fails for case classes with higher-kinded type parameters

### ISSUE THREAD

**Issue by mrdziuban** [Tinkoff/phobos#260]
Using scala 3, this code triggers an error at compile-time: https://scastie.scala-lang.org/mrdziuban/pGKWKNNeSCKQWVnM5mUJIw/1

```scala
import ru.tinkoff.phobos.decoding.XmlDecoder
import ru.tinkoff.phobos.derivation.semiauto.deriveXmlDecoder

case class Test[F[_]](fi: F[Int])

object Test {
  given xd: XmlDecoder[Test[Option]] = deriveXmlDecoder("Test")
}
```

```
Exception occurred while executing macro expansion.
java.lang.AssertionError: assertion failed: TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),HKTypeLambda(List(_$1), List(TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Any))), TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Any), List()))
  at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
  at dotty.tools.dotc.core.Types$TypeBounds.<init>(Types.scala:5104)
  at dotty.tools.dotc.core.Types$RealTypeBounds.<init>(Types.scala:5180)
  at dotty.tools.dotc.core.Types$TypeBounds$.apply(Types.scala:5224)
  at dotty.tools.dotc.core.Types$TypeBounds.derivedTypeBounds(Types.scala:5112)
  at dotty.tools.dotc.core.ConstraintHandling.addOneBound(ConstraintHandling.scala:267)
  at dotty.tools.dotc.core.ConstraintHandling.addOneBound$(ConstraintHandling.scala:27)
  at dotty.tools.dotc.core.ProperGadtConstraint.addOneBound(GadtConstraint.scala:60)
  at dotty.tools.dotc.core.ConstraintHandling.addBoundTransitively(ConstraintHandling.scala:321)
  at dotty.tools.dotc.core.ConstraintHandling.addBoundTransitively$(ConstraintHandling.scala:27)
  at dotty.tools.dotc.core.ProperGadtConstraint.addBoundTransitively(GadtConstraint.scala:60)
  at dotty.tools.dotc.core.ProperGadtConstraint.addBound(GadtConstraint.scala:167)
  at dotty.tools.dotc.core.TypeComparer.gadtAddLowerBound(TypeComparer.scala:118)
  at dotty.tools.dotc.core.TypeComparer.narrowGADTBounds(TypeComparer.scala:1931)
  at dotty.tools.dotc.core.TypeComparer.compareGADT$1(TypeComparer.scala:523)
  at dotty.tools.dotc.core.TypeComparer.thirdTryNamed$1(TypeComparer.scala:526)
  at dotty.tools.dotc.core.TypeComparer.thirdTry$1(TypeComparer.scala:575)
  at dotty.tools.dotc.core.TypeComparer.secondTry$1(TypeComparer.scala:506)
  at dotty.tools.dotc.core.TypeComparer.compareNamed$1(TypeComparer.scala:306)
  at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:312)
  at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1336)
  at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:194)
  at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:204)
  at dotty.tools.dotc.core.TypeComparer.topLevelSubType(TypeComparer.scala:128)
  at dotty.tools.dotc.core.TypeComparer$.topLevelSubType(TypeComparer.scala:2752)
  at dotty.tools.dotc.core.Types$Type.$less$colon$less(Types.scala:1044)
  at scala.quoted.runtime.impl.QuoteMatcher$.$eq$qmark$eq(QuoteMatcher.scala:345)
  at scala.quoted.runtime.impl.QuoteMatcher$.treeMatch(QuoteMatcher.scala:126)
  at scala.quoted.runtime.impl.QuotesImpl.scala$quoted$runtime$impl$QuotesImpl$$treeMatch(QuotesImpl.scala:3096)
  at scala.quoted.runtime.impl.QuotesImpl$TypeMatch$.unapply(QuotesImpl.scala:3066)
  at ru.tinkoff.phobos.derivation.decoder$.decodeElementCases$$anonfun$1(decoder.scala:113)
  at scala.collection.immutable.List.map(List.scala:246)
  at ru.tinkoff.phobos.derivation.decoder$.decodeElementCases(decoder.scala:130)
  at ru.tinkoff.phobos.derivation.decoder$.decodeStartElement(decoder.scala:141)
  at ru.tinkoff.phobos.derivation.decoder$.deriveProduct$$anonfun$1(decoder.scala:360)
  at ru.tinkoff.phobos.derivation.decoder$.deriveProduct$$anonfun$adapted$1(decoder.scala:370)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:109)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1452)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1511)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1432)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1434)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1511)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1432)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1440)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1513)
  at scala.collection.immutable.List.mapConserve(List.scala:472)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1513)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformSub(Trees.scala:1517)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1438)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1434)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1511)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1432)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1484)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1513)
  at scala.collection.immutable.List.mapConserve(List.scala:472)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1513)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1509)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1511)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1432)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1484)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1513)
  at scala.collection.immutable.List.mapConserve(List.scala:472)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1513)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1509)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1488)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1486)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1513)
  at scala.collection.immutable.List.mapConserve(List.scala:472)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1513)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1509)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1511)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1432)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.quoted.PickledQuotes$.spliceTerms(PickledQuotes.scala:151)
  at dotty.tools.dotc.quoted.PickledQuotes$.unpickleTerm(PickledQuotes.scala:87)
  at scala.quoted.runtime.impl.QuotesImpl.unpickleExprV2(QuotesImpl.scala:3044)
  at ru.tinkoff.phobos.derivation.decoder$.deriveProduct(decoder.scala:370)
  at ru.tinkoff.phobos.derivation.decoder$.deriveElementDecoderImpl(decoder.scala:39)
  at ru.tinkoff.phobos.derivation.decoder$.deriveXmlDecoderImpl$$anonfun$1(decoder.scala:52)
  at ru.tinkoff.phobos.derivation.decoder$.deriveXmlDecoderImpl$$anonfun$adapted$1(decoder.scala:52)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:109)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1452)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1513)
  at scala.collection.immutable.List.mapConserve(List.scala:472)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1513)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1418)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.quoted.PickledQuotes$.spliceTerms(PickledQuotes.scala:151)
  at dotty.tools.dotc.quoted.PickledQuotes$.unpickleTerm(PickledQuotes.scala:87)
  at scala.quoted.runtime.impl.QuotesImpl.unpickleExprV2(QuotesImpl.scala:3044)
  at ru.tinkoff.phobos.derivation.decoder$.deriveXmlDecoderImpl(decoder.scala:52)
```

**Comment by valentiay** [comment]
Hi!
Thanks for the bug report. It seems, that this error is caused by [this](https://github.com/Tinkoff/phobos/blob/master/modules/core/src/main/scala-3/ru/tinkoff/phobos/derivation/decoder.scala#L113) pattern matching. Unfortunately, I could not fix it yet nor did I found any workarounds. I guess, I'll have to rewrite this code without quotes and splices. It'll take some time, but I hope to finish it next week

**Comment by valentiay** [comment]
Unfortunately, I haven't yet managed to fix this bug. Things are more complicated than I thought, so I'll need some extra time

**Comment by valentiay** [comment]
Hi!
I've finnaly managed to fix this bug. The fix is published in v0.19.2

**Comment by mrdziuban** [comment]
Thank you!
